File: web-simulator/src/models/Player.js
// Player model - mirrors Swift Player @Model
export class Player {
  constructor(name) {
    this.id = crypto.randomUUID();
    this.name = name;
    this.totalGames = 0;
    this.totalScore = 0.0;
    this.firstPlaceCount = 0;
    this.secondPlaceCount = 0;
    this.thirdPlaceCount = 0;
    this.fourthPlaceCount = 0;
  }

  get averageRank() {
    if (this.totalGames === 0) return 0.0;
    const sum = (this.firstPlaceCount * 1) + (this.secondPlaceCount * 2) +
                (this.thirdPlaceCount * 3) + (this.fourthPlaceCount * 4);
    return sum / this.totalGames;
  }

  get top2Rate() {
    if (this.totalGames === 0) return 0.0;
    return (this.firstPlaceCount + this.secondPlaceCount) / this.totalGames;
  }

  get lastPlaceAvoidanceRate() {
    if (this.totalGames === 0) return 0.0;
    return (this.totalGames - this.fourthPlaceCount) / this.totalGames;
  }
}

File: swift/MahjongScoreApp/Models/Player.swift
import SwiftData
import Foundation

@Model
final class Player {
    var id: UUID = UUID()
    var name: String = ""
    
    // Stats
    var totalGames: Int = 0
    var totalScore: Double = 0.0
    var firstPlaceCount: Int = 0
    var secondPlaceCount: Int = 0
    var thirdPlaceCount: Int = 0
    var fourthPlaceCount: Int = 0
    
    // Computed Properties
    var averageRank: Double {
        guard totalGames > 0 else { return 0.0 }
        let sum = (firstPlaceCount * 1) + (secondPlaceCount * 2) + (thirdPlaceCount * 3) + (fourthPlaceCount * 4)
        return Double(sum) / Double(totalGames)
    }
    
    var top2Rate: Double {
        guard totalGames > 0 else { return 0.0 }
        return Double(firstPlaceCount + secondPlaceCount) / Double(totalGames)
    }
    
    var lastPlaceAvoidanceRate: Double {
        guard totalGames > 0 else { return 0.0 }
        return Double(totalGames - fourthPlaceCount) / Double(totalGames)
    }

    init(name: String) {
        self.id = UUID()
        self.name = name
    }
}
